File: src/types/video_speed.ts
/**
 * Mirror of `src-tauri/src/models/video_speed.rs`
 * (Calculador de Velocidade — Fase 2, camada de dados).
 *
 * Field names stay snake_case so serde defaults line up with the wire. These
 * interfaces describe the **structured** shape a future Tauri command returns —
 * the Rust repository (de)serializes the `*_json` table columns into these
 * types, so the front-end never sees the raw JSON-text columns.
 */

export type VideoSpeedMethod = "line" | "plane" | "cross_ratio";
export type VideoSpeedReferenceSource = "campo" | "norma_viaria" | "entre_eixos";

/** One pixel↔world correspondence used to fit the homography. */
export interface ControlPoint {
  px: number;
  py: number;
  world_x_m: number;
  world_y_m: number;
  label?: string | null;
}

export interface VideoSpeedCalibration {
  id: string;
  occurrence_id: string;
  /** sha256 of the source video. */
  media_hash: string;
  method: VideoSpeedMethod | string;
  control_points: ControlPoint[];
  reference_source: VideoSpeedReferenceSource | string;
  /** Row-major 3×3 homography (image px → world m), length 9. */
  homography: number[];
  /** RMS reprojection residual in pixels (null if not computed). */
  residuals_px: number | null;
  /** Reserved for a future lens-distortion model; null today. */
  distortion_model: Record<string, unknown> | null;
  author: string;
  created_at: string;
}

/**
 * One marked vehicle position, bound to a real collected storyboard frame so
 * it inherits that frame's timestamp (pericial reproducibility).
 */
export interface TrajectoryPoint {
  /** Storyboard frame this point was marked on (the collected frame). */
  storyboard_frame_id?: string | null;
  /** The export (PNG) backing the storyboard frame, when available. */
  export_id?: string | null;
  px: number;
  py: number;
  /** Pixel marking uncertainty (1σ) for this point. */
  u_px: number;
  /** Real frame time inherited from the storyboard frame. */
  actual_timestamp_s: number;
  /** ffmpeg seek error (requested − actual), inherited for audit. */
  delta_s?: number | null;
  /** True when marked by a human (always true in this phase). */
  manual: boolean;
}

/** Monte Carlo per-source sigmas — persisted so the run is reproducible. */
export interface McSigmas {
  calibration_px: number;
  world_m: number;
  trajectory_px: number;
  time_s: number;
}

/**
 * A speed calculation. Uncertainty fields are nullable because not every
 * calculation has them:
 *   - 2 points (average): velocity_kmh/vx/vy present, but NO CI and NO Monte
 *     Carlo — every nullable field is null, and `residuals` is empty.
 *   - ≥3 points with a plane (4-pt) calibration: regression (CI) + Monte Carlo.
 *   - ≥3 points with a line (2-pt) calibration: regression (CI) present, but
 *     Monte Carlo null (MC needs 4 coplanar points).
 */
export interface VideoSpeedCalculation {
  id: string;
  occurrence_id: string;
  media_hash: string;
  /** FK → VideoSpeedCalibration.id */
  calibration_id: string;
  points: TrajectoryPoint[];
  /** |v| = sqrt(vx²+vy²), in km/h (the headline value). */
  velocity_kmh: number;
  vx_m_per_s: number;
  vy_m_per_s: number;
  /** Standard error of |v| in m/s (null for the 2-point case). */
  se_m_per_s: number | null;
  /** Lower confidence-interval bound, in km/h (null for the 2-point case). */
  ci_low: number | null;
  /** Upper confidence-interval bound, in km/h (null for the 2-point case). */
  ci_high: number | null;
  /** Confidence level for [ci_low, ci_high], e.g. 0.95 (null if 2 points). */
  confidence: number | null;
  r_squared: number | null;
  /** 2D residual per point (empty for the 2-point case). */
  residuals: number[];
  /** Exact RNG seed used by the Monte Carlo run (null if MC did not run). */
  mc_seed: number | null;
  /** Exact per-source sigmas used (null if MC did not run). */
  mc_sigmas: McSigmas | null;
  /** Monte Carlo iterations requested (null if MC did not run). */
  mc_n: number | null;
  /** Iterations that failed and were discarded (null if MC did not run). */
  mc_failed: number | null;
  mc_mean_kmh: number | null;
  mc_median_kmh: number | null;
  mc_p2_5_kmh: number | null;
  mc_p97_5_kmh: number | null;
  /** Technical caveats to transcribe into the laudo. */
  limitations: string[];
  /** Free-form audit trail (tool versions, operator notes, etc.). */
  audit: Record<string, unknown> | null;
  author: string;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Command inputs (Phase 3)

export interface CreateSpeedCalibrationInput {
  /** sha256 of a video already registered in this occurrence. */
  media_hash: string;
  /** "line" (2 points) | "plane" (4 points, DLT). */
  method: VideoSpeedMethod | string;
  /** Pixel↔world correspondences (4 for plane, 2 for line). */
  control_points: ControlPoint[];
  reference_source: VideoSpeedReferenceSource | string;
  author?: string | null;
}

export interface ComputeSpeedInput {
  /** Stored calibration used to project pixel→world. */
  calibration_id: string;
  /** Marked trajectory — each point tied to a real collected frame. */
  points: TrajectoryPoint[];
  /** Monte Carlo iterations (>= 10). Ignored for the 2-point case. */
  mc_n?: number | null;
  /** Per-source sigmas for Monte Carlo. Ignored for the 2-point case. */
  mc_sigmas?: McSigmas | null;
  /** Desired CI level (only 0.95 is supported in this phase). */
  confidence?: number | null;
  author?: string | null;
}

File: src-tauri/src/models/video_speed.rs
//! Calculador de Velocidade — modelos persistidos (Fase 2: camada de dados).
//!
//! Espelham `src/types/video_speed.ts` no front-end. Os nomes de campo são
//! snake_case para casar com o wire serde. Os structs carregam tipos
//! **estruturados** (matriz da homografia, pontos da trajetória, sigmas do
//! Monte Carlo); o repositório (`video_speed_repo`) é quem (de)serializa
//! esses tipos para as colunas `*_json` da tabela.
//!
//! ## Reprodutibilidade pericial
//!
//! O número do laudo precisa ser reproduzível. Por isso:
//!   - [`VideoSpeedCalculation::mc_seed`] e [`VideoSpeedCalculation::mc_sigmas`]
//!     guardam **exatamente** o que o Monte Carlo usou.
//!   - Cada [`TrajectoryPoint`] aponta para um frame coletado real do
//!     storyboard (`storyboard_frame_id` / `export_id`), herdando
//!     `actual_timestamp_s` e `delta_s` — o tempo nunca é "inventado".

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

// ---------------------------------------------------------------------------
// Calibração

/// Uma correspondência pixel↔mundo usada para ajustar a homografia.
///
/// Para `method = "plane"` (DLT) há 4 pontos; para `method = "line"` há 2
/// (as duas pontas de um segmento cujo comprimento real está embutido nas
/// coordenadas de mundo).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct ControlPoint {
    pub px: f64,
    pub py: f64,
    pub world_x_m: f64,
    pub world_y_m: f64,
    #[serde(default)]
    pub label: Option<String>,
}

/// Calibração congelada: mapeia pixels da imagem para metros no plano da via.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct VideoSpeedCalibration {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    /// sha256 do vídeo de origem.
    pub media_hash: String,
    /// `"line"` | `"plane"`.
    pub method: String,
    pub control_points: Vec<ControlPoint>,
    /// `"campo"` | `"norma_viaria"` | `"entre_eixos"`.
    pub reference_source: String,
    /// Homografia 3x3 row-major (imagem px → mundo m), 9 f64.
    pub homography: [f64; 9],
    /// RMS de reprojeção em pixels (None se não calculado).
    pub residuals_px: Option<f64>,
    /// Reservado para um futuro modelo de distorção de lente; NULL hoje.
    #[serde(default)]
    pub distortion_model: Option<serde_json::Value>,
    pub author: String,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// Cálculo

/// Uma posição marcada do veículo, amarrada a um frame coletado real para
/// herdar o tempo do frame (reprodutibilidade pericial).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct TrajectoryPoint {
    /// Frame do storyboard onde o ponto foi marcado (o frame coletado).
    #[serde(default)]
    pub storyboard_frame_id: Option<Uuid>,
    /// O export (PNG) que respalda o frame do storyboard, quando disponível.
    #[serde(default)]
    pub export_id: Option<Uuid>,
    pub px: f64,
    pub py: f64,
    /// Incerteza de marcação em pixels (1σ) deste ponto.
    pub u_px: f64,
    /// Tempo real do frame, herdado do storyboard.
    pub actual_timestamp_s: f64,
    /// Erro de seek do ffmpeg (pedido − real), herdado para auditoria.
    #[serde(default)]
    pub delta_s: Option<f64>,
    /// True quando marcado por um humano (sempre true nesta fase).
    pub manual: bool,
}

/// Sigmas por fonte usados no Monte Carlo — persistidos para reprodutibilidade.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct McSigmas {
    pub calibration_px: f64,
    pub world_m: f64,
    pub trajectory_px: f64,
    pub time_s: f64,
}

/// Resultado de um cálculo de velocidade: ajuste por regressão + distribuição
/// Monte Carlo, com tudo o que o laudo precisa para reproduzir o número.
///
/// Os campos de incerteza são `Option` porque nem todo cálculo os tem:
///   - **2 pontos** (média): há `velocity_kmh`/`vx`/`vy`, mas SEM IC e SEM
///     Monte Carlo — todos os `Option` ficam `None`, e `residuals` vazio.
///   - **≥3 pontos, calibração de plano (4 pts) E σ informados pelo perito**:
///     regressão (IC) + Monte Carlo, todos preenchidos.
///   - **≥3 pontos SEM σ, ou calibração por linha (2 pts)**: regressão (IC)
///     preenchida, mas Monte Carlo `None` (MC exige 4 pontos coplanares e
///     incertezas informadas — não inventamos σ).
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct VideoSpeedCalculation {
    pub id: Uuid,
    pub occurrence_id: Uuid,
    pub media_hash: String,
    /// FK → [`VideoSpeedCalibration::id`].
    pub calibration_id: Uuid,
    pub points: Vec<TrajectoryPoint>,
    /// |v| = sqrt(vx²+vy²), em km/h (valor de destaque do laudo).
    pub velocity_kmh: f64,
    pub vx_m_per_s: f64,
    pub vy_m_per_s: f64,
    /// Erro-padrão de |v| em m/s (None no caso de 2 pontos).
    pub se_m_per_s: Option<f64>,
    /// Limite inferior do intervalo de confiança, em km/h (None se 2 pontos).
    pub ci_low: Option<f64>,
    /// Limite superior do intervalo de confiança, em km/h (None se 2 pontos).
    pub ci_high: Option<f64>,
    /// Nível de confiança de `[ci_low, ci_high]`, ex.: 0.95 (None se 2 pontos).
    pub confidence: Option<f64>,
    pub r_squared: Option<f64>,
    /// Resíduo 2D por ponto (vazio no caso de 2 pontos).
    pub residuals: Vec<f64>,
    /// Semente RNG resolvida usada pelo Monte Carlo (reprodutibilidade).
    /// QUANDO o MC roda, é SEMPRE persistida. None se o MC não rodou.
    pub mc_seed: Option<i64>,
    /// Sigmas por fonte exatos usados (reprodutibilidade). None se sem MC.
    pub mc_sigmas: Option<McSigmas>,
    /// Iterações Monte Carlo pedidas (None se sem MC).
    pub mc_n: Option<i64>,
    /// Iterações descartadas (ex.: homografia singular). None se sem MC.
    pub mc_failed: Option<i64>,
    pub mc_mean_kmh: Option<f64>,
    pub mc_median_kmh: Option<f64>,
    pub mc_p2_5_kmh: Option<f64>,
    pub mc_p97_5_kmh: Option<f64>,
    /// Ressalvas técnicas a transcrever no laudo.
    pub limitations: Vec<String>,
    /// Trilha de auditoria livre (versões de ferramentas, notas do operador…).
    #[serde(default)]
    pub audit: serde_json::Value,
    pub author: String,
    pub created_at: DateTime<Utc>,
}

// ---------------------------------------------------------------------------
// Inputs dos comandos Tauri (Fase 3)

/// Entrada de `create_speed_calibration`.
#[derive(Debug, Clone, Deserialize)]
pub struct CreateSpeedCalibrationInput {
    /// sha256 do vídeo já registrado nesta ocorrência.
    pub media_hash: String,
    /// `"line"` (2 pontos) | `"plane"` (4 pontos, DLT).
    pub method: String,
    /// Correspondências pixel↔mundo (4 para plane, 2 para line).
    pub control_points: Vec<ControlPoint>,
    /// `"campo"` | `"norma_viaria"` | `"entre_eixos"`.
    pub reference_source: String,
    /// Autor da calibração (opcional; default vazio).
    #[serde(default)]
    pub author: Option<String>,
}

/// Entrada de `compute_speed`.
#[derive(Debug, Clone, Deserialize)]
pub struct ComputeSpeedInput {
    /// Calibração já gravada a usar na projeção pixel→mundo.
    pub calibration_id: Uuid,
    /// Trajetória marcada — cada ponto amarrado a um frame coletado real.
    pub points: Vec<TrajectoryPoint>,
    /// Iterações Monte Carlo (>= 10). Ignorado no caso de 2 pontos.
    #[serde(default)]
    pub mc_n: Option<u32>,
    /// Sigmas por fonte para o Monte Carlo. Ignorado no caso de 2 pontos.
    #[serde(default)]
    pub mc_sigmas: Option<McSigmas>,
    /// Nível de confiança desejado para o IC (apenas 0.95 suportado nesta fase).
    #[serde(default)]
    pub confidence: Option<f64>,
    /// Autor do cálculo (opcional; default vazio).
    #[serde(default)]
    pub author: Option<String>,
}
